bokeh/plotting/contour.py

```diff
@@ -0,0 +1,278 @@
+# Copyright (c) 2012 - 2022, Anaconda, Inc., and Bokeh Contributors.
+# All rights reserved.
+#
+# The full license is in the file LICENSE.txt, distributed with this software.
+#-----------------------------------------------------------------------------
+
+#-----------------------------------------------------------------------------
+# Boilerplate
+#-----------------------------------------------------------------------------
+from __future__ import annotations
+
+import logging # isort:skip
+log = logging.getLogger(__name__)
+
+#-----------------------------------------------------------------------------
+# Imports
+#-----------------------------------------------------------------------------
+
+# Standard library imports
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Dict,
+    Optional,
+    Sequence,
+    Tuple,
+    Union,
+)
+
+# External imports
+import numpy as np
+from numpy.typing import ArrayLike
+
+# Bokeh imports
+from ..core.property_mixins import FillProps, HatchProps, LineProps
+from ..models.contour_renderer import ContourRenderer
+from ..models.glyphs import MultiLine, MultiPolygons
+from ..models.sources import ColumnDataSource
+from ..palettes import linear_palette
+from ..plotting._renderer import _process_sequence_literals
+
+#-----------------------------------------------------------------------------
+# Globals and constants
+#-----------------------------------------------------------------------------
+
+__all__ = (
+    'contour_data',
+    'from_contour',
+)
+
+#-----------------------------------------------------------------------------
+# General API
+#-----------------------------------------------------------------------------
+
+def contour_data(
+    x: ArrayLike | None = None,
+    y: ArrayLike | None = None,
+    z: ArrayLike | np.ma.MaskedArray | None = None,
+    levels: Sequence[float] = [],
+    *,
+    want_fill: bool = True,
+    want_line: bool = True,
+) -> Tuple[Optional[Dict[str, Any]], Optional[Dict[str, Any]]]:
+    '''
+    Return the contour data of filled and/or line contours that can be used to set
+    ContourRenderer.data
+    '''
+    levels = _validate_levels(levels)
+    if len(levels) < 2:
+        want_fill = False
+
+    if not want_fill and not want_line:
+        raise RuntimeError("Neither fill nor line requested in contour_data")
+
+    fill_coords, line_coords = _contour_coords(x, y, z, levels, want_fill, want_line)
+
+    fill_data = None
+    if fill_coords:
+        xs, ys = fill_coords
+        fill_data = dict(xs=xs, ys=ys, lower_levels=levels[:-1], upper_levels=levels[1:])
+
+    line_data = None
+    if line_coords:
+        xs, ys = line_coords
+        line_data = dict(xs=xs, ys=ys, levels=levels)
+
+    return fill_data, line_data
+
+#-----------------------------------------------------------------------------
+# Dev API
+#-----------------------------------------------------------------------------
+
+def from_contour(
```

Comment:
This is only ever called from `figure.contour`. It is public as far as this file is concerned, but should be considered private for bokeh as a whole. So maybe it should have a prefixed underscore?